lib/kadm5/init_c.c

```diff
@@ -106,42 +106,84 @@ _kadm5_c_init_context(kadm5_client_context **ctx,
 	free(*ctx);
 	return ret;
     }
-    if(params->mask & KADM5_CONFIG_ADMIN_SERVER)
+
+    /*
+     * FIXME: If we have a hostlist, we should use the hostlist so that if we
+     *        can't reach one server we try another.
+     */
+    if (params->mask & KADM5_CONFIG_ADMIN_SERVER)
 	(*ctx)->admin_server = strdup(params->admin_server);
     else {
 	char **hostlist;
 
-	ret = krb5_get_krb_admin_hst (context, &(*ctx)->realm, &hostlist);
+	ret = krb5_get_krb_admin_hst(context, &(*ctx)->realm, &hostlist);
 	if (ret) {
 	    free((*ctx)->realm);
 	    free(*ctx);
 	    return ret;
 	}
 	(*ctx)->admin_server = strdup(*hostlist);
-	krb5_free_krbhst (context, hostlist);
+	krb5_free_krbhst(context, hostlist);
     }
 
     if ((*ctx)->admin_server == NULL) {
 	free((*ctx)->realm);
 	free(*ctx);
 	return krb5_enomem(context);
     }
-    colon = strchr ((*ctx)->admin_server, ':');
+    colon = strchr((*ctx)->admin_server, ':');
     if (colon != NULL)
 	*colon++ = '\0';
 
     (*ctx)->kadmind_port = 0;
 
-    if(params->mask & KADM5_CONFIG_KADMIND_PORT)
+    if (params->mask & KADM5_CONFIG_KADMIND_PORT)
 	(*ctx)->kadmind_port = params->kadmind_port;
     else if (colon != NULL) {
 	char *end;
 
-	(*ctx)->kadmind_port = htons(strtol (colon, &end, 0));
+	(*ctx)->kadmind_port = htons(strtol(colon, &end, 0));
     }
     if ((*ctx)->kadmind_port == 0)
-	(*ctx)->kadmind_port = krb5_getportbyname (context, "kerberos-adm",
-						   "tcp", 749);
+	(*ctx)->kadmind_port = krb5_getportbyname(context, "kerberos-adm",
+						  "tcp", 749);
+
+    if (params->mask & KADM5_CONFIG_READONLY_ADMIN_SERVER)
+	(*ctx)->readonly_admin_server = strdup(params->readonly_admin_server);
+    else {
+	char **hostlist;
+
+        ret = krb5_get_krb_readonly_admin_hst(context, &(*ctx)->realm,
+                                              &hostlist);
+	if (ret) {
+	    free((*ctx)->realm);
+	    free(*ctx);
+	    return ret;
+	}
+	(*ctx)->readonly_admin_server = strdup(*hostlist);
+	krb5_free_krbhst(context, hostlist);
+    }
+
+    if ((*ctx)->readonly_admin_server == NULL) {
```

Comment:
Same as above, but this is worse because initially we'll never find read-only servers, so we'll always be failing.  BUG.